### ISSUE TITLE
Reading strings with low-level python API

### ISSUE THREAD

**Issue by isosc** [ornladios/ADIOS2#2338]
Are there any examples of reading strings with the low level API in python? I have only found this one: https://github.com/ornladios/ADIOS2/blob/master/testing/adios2/bindings/python/TestBPWriteReadTypes.py which suggests that the value of the string should be in the structure that is returned from Engine.AvailableVariables, however I don't see it there in practice. I also see a form of Engine.Get that takes a string, but I haven't been able to make that work. Is there a test or an example that illustrates best practice for string variables?

Thanks

**Comment by pnorbert** [comment]
@williamfgc Do you know how to read strings with python?

**Comment by isosc** [comment]
Okay, here's the code that I think is intended to work:
```python
from unittest import TestCase
import adios2


class TestAdiosFunctionality(TestCase):
    def test_adios_read_string_ll(self):
        the_string = 'hello adios'
        bpfilename = 'string_test.bp'
        varname = 'mystringvar'

        # test writing
        adios = adios2.ADIOS()
        ioWrite = adios.DeclareIO('ioWriter')
        ad_engine = ioWrite.Open(bpfilename, adios2.Mode.Write)
        ad_engine.BeginStep()
        var_mystringvar = ioWrite.DefineVariable(varname)
        ad_engine.Put(var_mystringvar, the_string)
        ad_engine.PerformPuts()
        ad_engine.EndStep()
        ad_engine.Close()

        ioRead = adios.DeclareIO('ioReader')
        ad_engine = ioRead.Open(bpfilename, adios2.Mode.Read)
        ad_engine.BeginStep()
        var_read_mystringvar = ioRead.InquireVariable(varname)
        result = ''
        ad_engine.Get(var_read_mystringvar, result)
        ad_engine.PerformGets()
        ad_engine.EndStep()
        ad_engine.Close()

        self.assertEqual(result, the_string)

```

But it fails at the assertEqual, with result==''. I think this is because in Python, strings are immutable, so creating a new string inside the Get call will result in an entirely different string being created, and thus won't have an impact on the result variable outside the Get call. So, the read is probably occurring properly, it just won't get passed back through this mechanism.

Perhaps I'm missing something, is there another way to read strings with the low-level python API?

**Comment by williamfgc** [comment]
@pnorbert I think @isosc just came across an old bug #1135 . I never found a good way to reallocate strings coming from Python on the C++ side using `pybind11` at that time (this is from almost 1.5 years ago, so things might have changed in pybind11). If possible, use the high-level APIs as that one creates and allocates the string in C++ as passes it to Python. Hope this helps.

**Comment by dmitry-ganyushin** [comment]
As a work-around one can use an array of integers like ord("a") = 97 etc..

**Comment by dmitry-ganyushin** [comment]
Fixed with PR #2489
